framework/src/main/java/org/checkerframework/common/accumulation/AccumulationAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,296 @@
+        return builder.build();
+    }
+
+    /**
+     * Returns true if the return type of the given method invocation tree has an @This annotation
+     * from the Returns Receiver Checker.
+     *
+     * @param tree a method invocation tree
+     * @return true if the method being invoked returns its receiver
+     */
+    public boolean returnsThis(final MethodInvocationTree tree) {
+        ReturnsReceiverAnnotatedTypeFactory rrATF =
+                getTypeFactoryOfSubchecker(ReturnsReceiverChecker.class);
+        ExecutableElement methodEle = TreeUtils.elementFromUse(tree);
+        AnnotatedTypeMirror methodAtm = rrATF.getAnnotatedType(methodEle);
+        AnnotatedTypeMirror rrType =
+                ((AnnotatedTypeMirror.AnnotatedExecutableType) methodAtm).getReturnType();
+        return rrType != null && rrType.hasAnnotation(This.class);
+    }
+
+    /**
+     * Is the given annotation an accumulator annotation?
+     *
+     * @param anm an annotation mirror
+     * @return true if the annotation mirror is an instance of this factory's accumulator annotation
+     */
+    public boolean isAccumulatorAnnotation(AnnotationMirror anm) {
+        return AnnotationUtils.areSameByClass(anm, ACC);
+    }
+
+    @Override
+    protected TreeAnnotator createTreeAnnotator() {
+        return new ListTreeAnnotator(
+                super.createTreeAnnotator(), new AccumulationTreeAnnotator(this));
+    }
+
+    /** Handles fluent APIs using the Returns Receiver Checker. */
+    protected class AccumulationTreeAnnotator extends TreeAnnotator {
+
+        /**
+         * Mandatory constructor.
+         *
+         * @param factory the type factory
+         */
+        public AccumulationTreeAnnotator(AccumulationAnnotatedTypeFactory factory) {
+            super(factory);
+        }
+
+        /**
+         * If a method returns its receiver, and the receiver has an accumulation type, then the
+         * default type of its return value is the type of the receiver.
+         *
+         * @param tree a method invocation tree
+         * @param type the type of that tree (i.e. the return type of the invoked method). Is
+         *     (possibly) side-effected by this method.
+         * @return nothing, works by side-effect on {@code type}
+         */
+        @Override
+        public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror type) {
+            if (returnsThis(tree)) {
+                // There is a @This annotation on the return type of the invoked method.
+
+                // Fetch the current type of the receiver, or top if none exists.
+                ExpressionTree receiverTree = TreeUtils.getReceiverTree(tree.getMethodSelect());
+                AnnotatedTypeMirror receiverType;
+                AnnotationMirror receiverAnno;
+
+                if (receiverTree != null
+                        && (receiverType = getAnnotatedType(receiverTree)) != null) {
```

Comment:
This particular code snippet is hard to read either way, and using an assignment within a conditional less verbose. Nevertheless, I've changed it to comply with the Checker Framework's typical code style, because an assignment within a conditional will be more confusing if they are not commonly used in a particular code base.